### ISSUE TITLE
Cursor jumps to first line when a hidden terminal buffer is created and a User event is executed after BufFilePost

### ISSUE THREAD

**Issue by bfrg** [vim/vim#5820]
**Describe the bug**
The cursor always jumps to the first line of the current buffer when a hidden terminal buffer is created and a `User` event is executed after `BufFilePost`.

This bug was introduced in `8.2.413`, see commit 5e94a29.

**To Reproduce**
1. Run `vim --clean tmp.vim -c "source %"` with
```vim
" tmp.vim
function On_BufFilePost()
    doautocmd <nomodeline> User UserEvent
endfunction

augroup test
    autocmd!
    autocmd BufFilePost * call On_BufFilePost()
    autocmd User UserEvent silent
augroup END
```
2. Move the cursor to the bottom of the buffer with `G`. Optionally scroll up with `zt` to make the bug more apparent.
3. Open a hidden terminal buffer (can also be in a popup window):
```vim
call term_start(&shell, #{term_finish: 'close', hidden: 1})
```
4. The cursor has jumped to the first line of the current buffer after the hidden terminal buffer has been created. The column number is always preserved though.

**Expected behavior**
The cursor should stay at its current position when a hidden terminal buffer is created.

**Environment**
- Vim version `8.2.422`
- OS: Gentoo Linux
- Terminal: xfce4-terminal, xterm, gvim

**Additional context**
I use the plugin [vim-projectionist](https://github.com/tpope/vim-projectionist) which executes a User event `ProjectionistDetect` after `BufFilePost`. Since `8.2.413` whenever I open a terminal buffer in a popup window, the cursor jumps to the first line of the buffer the popup terminal was opened from.

**Comment by brammool** [comment]
I cannot reproduce this.

**Comment by bfrg** [comment]
Here's an asciinema video that illustrates it: https://asciinema.org/a/7obYoazxsKSfwUFuXSwi73asx

**Comment by mcritchlow** [comment]
I can confirm/reproduce this issue w/ vim-projectionist as well.

**Comment by formationPythonPC** [comment]
I have the same problem and no solution (I have tried with markers, functions getpos, setpos but also with winsaveview, winrestview, the result is the same : cursor in "gg" after changing buffer.

- Vim version : 8.2

- OS : Archlinux

- Terminal : alacritty

**Comment by formationPythonPC** [comment]
I had the same problem and no solution (I have tried with markers, functions getpos, setpos but also with winsaveview, winrestview, the result is the same : cursor in "gg" after changing buffer.

Vim version : 8.2.1704

OS : Archlinux

Terminal : alacritty

====> Solved for me : it was due to the use of BufEnter in my vimrc.

**Comment by mmrwoods** [comment]
I've also run into this problem since upgrading my vim :-(

Although it is admittedly a minor issue, it is quite annoying and it would be great to see this fixed sometime soon.

I can reproduce the issue using the steps outlined by @bfrg, it does seem to be an issue with vim rather than anything specific to the plugins referenced within this GH issue.

On a possibly related note, when I tried to work around this issue within a `BufFilePost` autocmd, by doing nothing when `&buftype` for the autocmd buffer is set to "terminal", I realised that this could not work because the buffer type is set after triggering `BufFilePost`. 

See https://github.com/vim/vim/blob/5e94a29ebbde10dd973d58f1adba9a2fc83877d1/src/terminal.c#L576-L585

I'm not sure whether this is by design or not, just flagging it in case it was not intentional.
